python/ray/data/datasource/file_meta_provider.py

```diff
@@ -46,3 +53,271 @@ def __call__(
         **kwargs,
     ) -> BlockMetadata:
         return self._get_block_metadata(paths, schema, **kwargs)
+
+
+@DeveloperAPI
+class BaseFileMetadataProvider(FileMetadataProvider):
```

Comment:
Similarly for this class?

Reply:
I have the same slight preference to preserve the structure here for this class as above - namely that `BaseFileMetadataProvider` remain the developer API that end-users implement their own custom `FileBasedDatasource` metadata providers against, that `DefaultFileMetadataProvider` and `FastFileMetadataProvider` remain more hidden Ray Data implementation details, and that implementors think carefully about if they need path expansion (plus how they would best implement it if so, since a metadata provider specific to a given cloud service or data platform could be much more performant/scalable here) and how they should best retrieve block metadata for their use-case.

Reply:
In both cases, the other part of this pertains to breaking backwards compatibility, since any custom metadata providers that may already be inheriting from the existing `BaseFileMetadataProvider` and `ParquetMetadataProvider` classes would be broken by this change.